go-controller/pkg/ovn/pod_selector_address_set.go

```diff
@@ -0,0 +1,639 @@
+package ovn
+
+import (
+	"fmt"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	libovsdbclient "github.com/ovn-org/libovsdb/client"
+	"github.com/ovn-org/libovsdb/ovsdb"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/libovsdbops"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+	addressset "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/address_set"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	kapi "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	kerrorsutil "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+)
+
+type PodSelectorAddressSet struct {
+	controllerName string
+	// PodSelectorAddressSet methods are called from network policy handlers, and local handlers asynchronously.
+	// To make sure local handlers won't try to update deleted resources, this lock is used together with deleted field.
+	sync.RWMutex
+	// this is a signal for local event handlers that they are/should be stopped.
+	// it will be set to true before any PodSelectorAddressSet infrastructure is deleted,
+	// therefore every handler can either do its work and be sure all required resources are there,
+	// or this value will be set to true and handler can't proceed.
+	// Use PodSelectorAddressSet.RLock to read this field and hold it for the whole event handling.
+	deleted bool
+
+	// unique key that identifies given PodSelectorAddressSet
+	key        string
+	addressSet addressset.AddressSet
+
+	// backRefs is a map of objects that use this address set.
+	// keys must be unique for all possible users, e.g. for NetworkPolicy use (np *networkPolicy) getKeyWithKind()
+	backRefs map[string]bool
```

Comment:
backRefs looks like it is accessed without locking the struct itself. Should this be a syncmap? Should it be Rlocked before it is read?

Reply:
actually it is always accessed with `oc.podSelectorAddressSets.DoWithLock`, since it is only changed on external calls, and handling may include create/delete the address set. I will add a comment for this field